test/lib/test_solve.ml

```diff
@@ -204,11 +204,71 @@ depends: ["p1" "p2"]
   calculate universe root
     [ ("ocaml-base-compiler", "3.14"); ("p1", "1"); ("p2", "1"); ("root", "0") ]
 
+let no_conflict_with_pin () =
+  let universe =
+    List.map OpamFile.OPAM.read_from_string
+      [
+        {|
+opam-version: "2.0"
+name: "ocaml-base-compiler"
+version: "3.14"
+|};
+        {|
+opam-version: "2.0"
+name: "p1"
+version: "1"
+dev-repo: "x"
+url {
+  src: "https://p.com/p.tbz"
+  checksum: "sha256=0000000000000000000000000000000000000000000000000000000000000001"
+}
+|};
+        {|
+opam-version: "2.0"
+name: "ocaml-base-compiler"
+version: "3.14"
+|};
+        {|
+opam-version: "2.0"
+name: "p1"
+version: "2"
+dev-repo: "x"
+url {
+  src: "https://p.com/p.tbz"
+  checksum: "sha256=0000000000000000000000000000000000000000000000000000000000000002"
+}
+|};
+        {|
+opam-version: "2.0"
+name: "p2"
+version: "dev"
+dev-repo: "x"
+url {
+  src: "git+https://x#hash"
+}
+|};
+      ]
+  in
+  let root =
+    OpamFile.OPAM.read_from_string
+      {|
+opam-version: "2.0"
+name: "root"
+version: "0"
+depends: ["p1" "p2"]
+|}
+  in
+  calculate universe root
```

Comment:
I must be missing something but how does this test verify anything about pinning?

Reply:
a pinned package in the opam universe is represented as a package with a single version and no hash. I unfortunately don't think there's a better way to detect this.

Reply:
It may be of interest to look at what I did in #353.

Reply:
I've pushed a better way to find out pinned packages. It's all a bit of a mess, so I'm not happy with that solution.

I agree we probably need to layer something like what @reynir has proposed to ensure the pinned packages specified in the root packages interact correctly with the pinned packages in the switch.